Codegen: TraceType only includes operators being registered

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #69840
* #70021
* #68462
* #68693
* #68692
* __->__ #68691

TraceType is a sharded file, so by only including specific operator
headers, we ensure that changing one (non-method) operator only needs
one shard to be re-compiled.

This also changes all the included autograd and jit headers from
including `ATen/ATen.h` to just including `ATen/core/Tensor.h`.

Differential Revision: [D33336948](https://our.internmc.facebook.com/intern/diff/D33336948)

peterbell10 committed:
Codegen: TraceType only includes operators being registered

TraceType is a sharded file, so by only including specific operator
headers, we ensure that changing one (non-method) operator only needs
one shard to be re-compiled.

This also changes all the included autograd and jit headers from
including `ATen/ATen.h` to just including `ATen/core/Tensor.h`.

[ghstack-poisoned]

peterbell10 committed:
Update on "Codegen: TraceType only includes operators being registered"


TraceType is a sharded file, so by only including specific operator
headers, we ensure that changing one (non-method) operator only needs
one shard to be re-compiled.

This also changes all the included autograd and jit headers from
including `ATen/ATen.h` to just including `ATen/core/Tensor.h`.

[ghstack-poisoned]

peterbell10 committed:
Update on "Codegen: TraceType only includes operators being registered"


TraceType is a sharded file, so by only including specific operator
headers, we ensure that changing one (non-method) operator only needs
one shard to be re-compiled.

This also changes all the included autograd and jit headers from
including `ATen/ATen.h` to just including `ATen/core/Tensor.h`.

[ghstack-poisoned]

peterbell10 committed:
Update on "Codegen: TraceType only includes operators being registered"


TraceType is a sharded file, so by only including specific operator
headers, we ensure that changing one (non-method) operator only needs
one shard to be re-compiled.

This also changes all the included autograd and jit headers from
including `ATen/ATen.h` to just including `ATen/core/Tensor.h`.

[ghstack-poisoned]

peterbell10 committed:
Update on "Codegen: TraceType only includes operators being registered"


TraceType is a sharded file, so by only including specific operator
headers, we ensure that changing one (non-method) operator only needs
one shard to be re-compiled.

This also changes all the included autograd and jit headers from
including `ATen/ATen.h` to just including `ATen/core/Tensor.h`.

[ghstack-poisoned]

peterbell10 committed:
Update on "Codegen: TraceType only includes operators being registered"


TraceType is a sharded file, so by only including specific operator
headers, we ensure that changing one (non-method) operator only needs
one shard to be re-compiled.

This also changes all the included autograd and jit headers from
including `ATen/ATen.h` to just including `ATen/core/Tensor.h`.

[ghstack-poisoned]

peterbell10 committed:
Update on "Codegen: TraceType only includes operators being registered"


TraceType is a sharded file, so by only including specific operator
headers, we ensure that changing one (non-method) operator only needs
one shard to be re-compiled.

This also changes all the included autograd and jit headers from
including `ATen/ATen.h` to just including `ATen/core/Tensor.h`.

[ghstack-poisoned]

albanD commented on tools/autograd/templates/TraceType.cpp:
```diff
@@ -1,17 +1,19 @@
+#define TORCH_ASSERT_ONLY_METHOD_OPERATORS
```
Do we want to put this macro more in the headers that were cleaned?

peterbell10 replied:
> Do we want to put this macro more in the headers that were cleaned?

It's sufficient for the header to be included by a file that defines the macro. Defining it in the header itself is possible, but you need to be very careful that it doesn't leak outside of the header. Otherwise, you might break the build of any source files that includes it. So, you have to guard it like this:

```
#pragma push_macro("TORCH_ASSERT_ONLY_METHOD_OPERATORS")
#define TORCH_ASSERT_ONLY_METHOD_OPERATORS

// rest of file

#pragma pop_macro("TORCH_ASSERT_ONLY_METHOD_OPERATORS")
```

Only defining it in source files is probably the safer route.

>Wondering about what should be the general policy around the forbid macro though?

Ideally this should be gradually introduced to all `.cpp` and `.cu` files until all of PyTorch can be compiled incrementally. At that point I would like to see it defined in the build system so all new files use `/ops` by default. Although you could explicitly opt-out by `undef`ing the macro so `derivatives.yaml`, for example, can still assume all operators are visible.

albanD commented on torch/csrc/utils/variadic.cpp:
```diff
@@ -1 +1,3 @@
 // Intentionally blank so I get a compile_commands.json entry for the header
+#define TORCH_ASSERT_ONLY_METHOD_OPERATORS
```
The comment above can be removed?

albanD reviewed (commented):
Sounds good. Wondering about what should be the general policy around the forbid macro though?

Also this will need a sister xla PR?

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32596264).

peterbell10 commented:
>Also this will need a sister xla PR?

Yes, pytorch/xla#3223 just adds a few missing includes. That was enough for xla to compile locally.

peterbell10 committed:
Update on "Codegen: TraceType only includes operators being registered"


TraceType is a sharded file, so by only including specific operator
headers, we ensure that changing one (non-method) operator only needs
one shard to be re-compiled.

This also changes all the included autograd and jit headers from
including `ATen/ATen.h` to just including `ATen/core/Tensor.h`.

Differential Revision: [D32596264](https://our.internmc.facebook.com/intern/diff/D32596264)

[ghstack-poisoned]

peterbell10 committed:
Update on "Codegen: TraceType only includes operators being registered"


TraceType is a sharded file, so by only including specific operator
headers, we ensure that changing one (non-method) operator only needs
one shard to be re-compiled.

This also changes all the included autograd and jit headers from
including `ATen/ATen.h` to just including `ATen/core/Tensor.h`.

Differential Revision: [D32596264](https://our.internmc.facebook.com/intern/diff/D32596264)

[ghstack-poisoned]

peterbell10 committed:
Update on "Codegen: TraceType only includes operators being registered"


TraceType is a sharded file, so by only including specific operator
headers, we ensure that changing one (non-method) operator only needs
one shard to be re-compiled.

This also changes all the included autograd and jit headers from
including `ATen/ATen.h` to just including `ATen/core/Tensor.h`.

Differential Revision: [D32596264](https://our.internmc.facebook.com/intern/diff/D32596264)

[ghstack-poisoned]

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32596264).

peterbell10 committed:
Rebase to fix merge conflicts on "Codegen: TraceType only includes operators being registered"


TraceType is a sharded file, so by only including specific operator
headers, we ensure that changing one (non-method) operator only needs
one shard to be re-compiled.

This also changes all the included autograd and jit headers from
including `ATen/ATen.h` to just including `ATen/core/Tensor.h`.

Differential Revision: [D32596264](https://our.internmc.facebook.com/intern/diff/D32596264)

[ghstack-poisoned]

peterbell10 committed:
Update on "Codegen: TraceType only includes operators being registered"


TraceType is a sharded file, so by only including specific operator
headers, we ensure that changing one (non-method) operator only needs
one shard to be re-compiled.

This also changes all the included autograd and jit headers from
including `ATen/ATen.h` to just including `ATen/core/Tensor.h`.

Differential Revision: [D32596264](https://our.internmc.facebook.com/intern/diff/D32596264)

[ghstack-poisoned]

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32596264).

peterbell10 committed:
Update on "Codegen: TraceType only includes operators being registered"


TraceType is a sharded file, so by only including specific operator
headers, we ensure that changing one (non-method) operator only needs
one shard to be re-compiled.

This also changes all the included autograd and jit headers from
including `ATen/ATen.h` to just including `ATen/core/Tensor.h`.

Differential Revision: [D32596264](https://our.internmc.facebook.com/intern/diff/D32596264)

[ghstack-poisoned]

peterbell10 committed:
Fix mypy issues on "Codegen: TraceType only includes operators being registered"


TraceType is a sharded file, so by only including specific operator
headers, we ensure that changing one (non-method) operator only needs
one shard to be re-compiled.

This also changes all the included autograd and jit headers from
including `ATen/ATen.h` to just including `ATen/core/Tensor.h`.

Differential Revision: [D32596264](https://our.internmc.facebook.com/intern/diff/D32596264)

[ghstack-poisoned]

peterbell10 committed:
Update on "Codegen: TraceType only includes operators being registered"


TraceType is a sharded file, so by only including specific operator
headers, we ensure that changing one (non-method) operator only needs
one shard to be re-compiled.

This also changes all the included autograd and jit headers from
including `ATen/ATen.h` to just including `ATen/core/Tensor.h`.

Differential Revision: [D32596264](https://our.internmc.facebook.com/intern/diff/D32596264)

[ghstack-poisoned]

peterbell10 committed:
Update on "Codegen: TraceType only includes operators being registered"


TraceType is a sharded file, so by only including specific operator
headers, we ensure that changing one (non-method) operator only needs
one shard to be re-compiled.

This also changes all the included autograd and jit headers from
including `ATen/ATen.h` to just including `ATen/core/Tensor.h`.

Differential Revision: [D32596264](https://our.internmc.facebook.com/intern/diff/D32596264)

[ghstack-poisoned]

peterbell10 committed:
Commit to re-run CI on "Codegen: TraceType only includes operators being registered"


TraceType is a sharded file, so by only including specific operator
headers, we ensure that changing one (non-method) operator only needs
one shard to be re-compiled.

This also changes all the included autograd and jit headers from
including `ATen/ATen.h` to just including `ATen/core/Tensor.h`.

Differential Revision: [D32596264](https://our.internmc.facebook.com/intern/diff/D32596264)

[ghstack-poisoned]

peterbell10 committed:
Update on "Codegen: TraceType only includes operators being registered"


TraceType is a sharded file, so by only including specific operator
headers, we ensure that changing one (non-method) operator only needs
one shard to be re-compiled.

This also changes all the included autograd and jit headers from
including `ATen/ATen.h` to just including `ATen/core/Tensor.h`.

Differential Revision: [D32596264](https://our.internmc.facebook.com/intern/diff/D32596264)

[ghstack-poisoned]

peterbell10 committed:
Update on "Codegen: TraceType only includes operators being registered"


TraceType is a sharded file, so by only including specific operator
headers, we ensure that changing one (non-method) operator only needs
one shard to be re-compiled.

This also changes all the included autograd and jit headers from
including `ATen/ATen.h` to just including `ATen/core/Tensor.h`.

Differential Revision: [D32596264](https://our.internmc.facebook.com/intern/diff/D32596264)

[ghstack-poisoned]

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32596264).

peterbell10 committed:
Update on "Codegen: TraceType only includes operators being registered"


TraceType is a sharded file, so by only including specific operator
headers, we ensure that changing one (non-method) operator only needs
one shard to be re-compiled.

This also changes all the included autograd and jit headers from
including `ATen/ATen.h` to just including `ATen/core/Tensor.h`.

Differential Revision: [D32596264](https://our.internmc.facebook.com/intern/diff/D32596264)

[ghstack-poisoned]

peterbell10 committed:
Rebase to fix merge conflicts on "Codegen: TraceType only includes operators being registered"


TraceType is a sharded file, so by only including specific operator
headers, we ensure that changing one (non-method) operator only needs
one shard to be re-compiled.

This also changes all the included autograd and jit headers from
including `ATen/ATen.h` to just including `ATen/core/Tensor.h`.

Differential Revision: [D32596264](https://our.internmc.facebook.com/intern/diff/D32596264)

[ghstack-poisoned]

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32596264).

albanD commented:
This one will take a bit longer. A lot of internal code relied on these headers importanting all the ops. I'll just fix them to do the proper imports.

albanD commented:
Ho no, my rebasing was wrong... Sorry about that.

peterbell10 committed:
Update on "Codegen: TraceType only includes operators being registered"


TraceType is a sharded file, so by only including specific operator
headers, we ensure that changing one (non-method) operator only needs
one shard to be re-compiled.

This also changes all the included autograd and jit headers from
including `ATen/ATen.h` to just including `ATen/core/Tensor.h`.

Differential Revision: [D32596264](https://our.internmc.facebook.com/intern/diff/D32596264)

[ghstack-poisoned]

peterbell10 commented:
@albanD rebased and fixed the merge conflicts.

albanD commented:
Thanks!

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33217915).

peterbell10 committed:
Update on "Codegen: TraceType only includes operators being registered"


TraceType is a sharded file, so by only including specific operator
headers, we ensure that changing one (non-method) operator only needs
one shard to be re-compiled.

This also changes all the included autograd and jit headers from
including `ATen/ATen.h` to just including `ATen/core/Tensor.h`.

Differential Revision: [D33217915](https://our.internmc.facebook.com/intern/diff/D33217915)

[ghstack-poisoned]

peterbell10 commented:
@pytorchbot ciflow rerun

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33217915).

albanD commented:
The latest rebase is breaking a few things:
- Internal caffe2 build fails with
```
caffe2/aten/src/ATen/core/aten_interned_strings.h:4:2: error: This change adds a dependency on native_functions.yaml,            meaning the file will need to be re-compiled every time an operator     is changed or added. Consider if including <ATen/core/symbol.h> for     the c10::Symbol class would be sufficient, or if your change would be   better placed in another file.
#error This change adds a dependency on native_functions.yaml,          \
```
- Breaks [glow](https://github.com/pytorch/glow) with
```
stderr: glow/glow/torch_glow/src/TorchGlowBackend.cpp:668:17: error: no type named 'EraseShapeInformation' in namespace 'torch::jit'
    torch::jit::EraseShapeInformation(graph);
    ~~~~~~~~~~~~^
```
It would be nice if we could avoid having to change glow in that same PR.

peterbell10 commented:
For the first one you need to look at the traceback and determine which cpp file is failing and which header are causing the issues.  If the cpp file isn't TraceType then maybe you can just remove the `TORCH_ASSERT_ONLY_METHOD_OPERATORS` otherwise you need to try removing the problematic includes.

For the `glow` issue, they are missing an include of `torch/csrc/jit/passes/shape_analysis.h`

peterbell10 commented:
@pytorchbot ciflow rerun -l ciflow/cpu

peterbell10 committed:
Update on "Codegen: TraceType only includes operators being registered"


TraceType is a sharded file, so by only including specific operator
headers, we ensure that changing one (non-method) operator only needs
one shard to be re-compiled.

This also changes all the included autograd and jit headers from
including `ATen/ATen.h` to just including `ATen/core/Tensor.h`.

Differential Revision: [D33217915](https://our.internmc.facebook.com/intern/diff/D33217915)

[ghstack-poisoned]

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33217915).

albanD commented:
I think the problem is that https://github.com/pytorch/pytorch/blob/b376d82cafe22ab18b563fff4b294e11402c46b8/aten/src/ATen/core/aten_interned_strings.h#L3 is missing the `AT_PER_OPERATOR_HEADERS` no?

peterbell10 commented:
That shouldn't be necessary because there is no per-operator header for `aten_interned_strings.h`. There must be an fb-internal change that's adding extra header dependencies that results in `aten_interned_strings.h` being included where it shouldn't.

albanD commented:
I don't think it is internal, the stack of include only includes pytorch files right:
```
stderr: In file included from buck-out/dev/gen/caffe2/generate-code/autograd/generated/TraceType_0.cpp:2:
In file included from caffe2/torch/csrc/autograd/VariableTypeUtils.h:12:
In file included from caffe2/torch/csrc/jit/frontend/tracer.h:9:
In file included from caffe2/torch/csrc/jit/api/object.h:6:
In file included from caffe2/torch/csrc/jit/api/method.h:7:
In file included from caffe2/torch/csrc/jit/api/function_impl.h:4:
In file included from caffe2/torch/csrc/jit/ir/ir.h:16:
In file included from caffe2/aten/src/ATen/core/interned_strings.h:10:
caffe2/aten/src/ATen/core/aten_interned_strings.h:4:2: error: This change adds a dependency on native_functions.yaml,            meaning the file will need to be re-compiled every time an operator     is changed or added. Consider if including <ATen/core/symbol.h> for     the c10::Symbol class would be sufficient, or if your change would be   better placed in another file.
```

peterbell10 commented:
@albanD if you look at the diff, this PR removes the includes for `VariableTypeUtils.h` in `TraceType.cpp` and the `object.h` include from `tracer.h`. So that include stack doesn't make sense unless they are coming from fb-internal code.

albanD commented:
Ho the changes to `VariableTypeUtils.h` are not on the internal diff. That is very weird.

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33336948).

albanD commented:
Ok after a clean import, it is much better, only the `benchmarks` are failing now:
```
stderr: In file included from caffe2/benchmarks/static_runtime/test_utils.cc:4:
In file included from caffe2/benchmarks/static_runtime/test_utils.h:8:
In file included from caffe2/torch/csrc/jit/ir/ir.h:16:
In file included from caffe2/aten/src/ATen/core/interned_strings.h:10:
caffe2/aten/src/ATen/core/aten_interned_strings.h:4:2: error: This change adds a dependency on native_functions.yaml,            meaning the file will need to be re-compiled every time an operator     is changed or added. Consider if including <ATen/core/symbol.h> for     the c10::Symbol class would be sufficient, or if your change would be   better placed in another file.
#error This change adds a dependency on native_functions.yaml,          \
 ```

Are the benchmarks tools from https://github.com/pytorch/pytorch/tree/master/benchmarks/static_runtime compiling properly for you?

peterbell10 committed:
Update on "Codegen: TraceType only includes operators being registered"


TraceType is a sharded file, so by only including specific operator
headers, we ensure that changing one (non-method) operator only needs
one shard to be re-compiled.

This also changes all the included autograd and jit headers from
including `ATen/ATen.h` to just including `ATen/core/Tensor.h`.

Differential Revision: [D33336948](https://our.internmc.facebook.com/intern/diff/D33336948)

[ghstack-poisoned]

peterbell10 commented:
Yeah, that looks real. Static runtime benchmarks aren't built by default so I didn't catch it locally.

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33336948).

albanD commented:
Thanks for the quick fix!
